vllm/distributed/communication_op.py

```diff
@@ -6,9 +6,38 @@
 from .parallel_state import get_tp_group
 
 
-def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
+@torch.library.custom_op("vllm::tp_out_of_place_ar",
+                         mutates_args=["input_"],
+                         device_types=("cuda", "cpu"))
+def _tp_out_of_place_ar(input_: torch.Tensor) -> torch.Tensor:
+    """All-reduce the input tensor across model parallel group."""
+    return get_tp_group().out_of_place_ar(input_)
+
+
+@torch.library.register_fake("vllm::tp_out_of_place_ar")
+def _tp_out_of_place_ar_fake(input_: torch.Tensor) -> torch.Tensor:
+    return input_
+
+
+@torch.library.custom_op("vllm::tp_in_place_ar",
+                         mutates_args=["input_"],
+                         device_types=("cuda", "cpu"))
+def _tp_in_place_ar(input_: torch.Tensor) -> None:
     """All-reduce the input tensor across model parallel group."""
-    return get_tp_group().all_reduce(input_)
+    get_tp_group().in_place_ar(input_)
+
+
+@torch.library.register_fake("vllm::tp_in_place_ar")
+def _tp_in_place_ar_fake(input_: torch.Tensor) -> None:
+    return
```

Comment:
I don't think nullary ops require meta-functions.

Reply:
Yeah, it looks like that's true.